File: lib/providers/DatabaseProvider.dart
import 'dart:io';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:user_management_tool/globals.dart';
import 'package:user_management_tool/models/Credentials.dart';
import 'package:yaml/yaml.dart';
import 'package:mongo_dart/mongo_dart.dart';
import 'package:user_management_tool/models/User.dart';

class DatabaseProvider {
  static var db, userCollection;

  // MongoDatabase({
  //   File configFile = File('pubspec.yaml');
  //   String yamlString = configFile.readAsStringSync();
  //   Map yaml = loadYaml(yamlString);
  // });

  static init() async {
    await connect();
    userCollection = db.collection("users");
    await insertUser(
      User(
        username: "admin",
        privileged: true,
        enabled: true,
      ),
    );
  }

  static connect() async {
    final dbUser = dotenv.get("MONGO_USER");
    final dbPassword = dotenv.get("MONGO_PASSWORD");
    final dbHost = dotenv.get("MONGO_HOST");
    final dbPort = dotenv.get("MONGO_PORT");
    final dbName = dotenv.get("MONGO_DATABASE");
    db = await Db.create(
      "mongodb://$dbUser:$dbPassword@$dbHost:$dbPort/$dbName",
    );
    await db.open();
  }

  static findUser(User user) async {
    var u = await userCollection.findOne({"username": user.username});
    if (u != null) {
      u = User.fromMap(u);
    }
    return u ?? false;
  }

  static checkCredentials(Credentials creds) async {
    var u = await userCollection.findOne({
      "username": creds.username,
      "password": creds.password,
    });
    if (u != null) {
      u = User.fromMap(u);
    }
    return u ?? false;
  }

  static Future<List<dynamic>> getUsers() async {
    var users = await userCollection.find().toList();
    if (users != null) {
      users = [for (var map in users) User.fromMap(map)];
    }
    return users ?? false;
  }

  static insertUser(User user) async {
    if (await findUser(user) == false) {
      await userCollection.insertOne(user.toMap());
      return true;
    } else {
      return false;
    }
  }

  static updateUser(User user) async {
    await userCollection.replaceOne({"username": user.username}, user.toMap());
    return true;
  }

  static deleteUser(User user) async {
    await userCollection.remove(where.eq('username', user.username));
    return true;
  }
}

File: providers/database_provider.py
import os
from pymongo import MongoClient
from models.credentials import Credentials
from models.user import User

from providers.dotenv_provider import CONFIG, ROOT_DIR


def connect():
    cs = f"mongodb://{CONFIG['MONGO_USER']}:{CONFIG['MONGO_PASSWORD']}@{CONFIG['MONGO_HOST']}:{CONFIG['MONGO_PORT']}"
    return MongoClient(cs)[CONFIG["MONGO_DATABASE"]]


def get_users():
    db = connect()
    users = db['users'].find()
    if (users != None):
        users = [User.fromJson(u) for u in users]
    return users


def find_user(username):
    assert isinstance(username, str), 'Argument of wrong type!'
    db = connect()
    user = db['users'].find_one({"username": username})
    if (user != None):
        user = User.fromJson(user)
    return user


def insert_user(user):
    assert isinstance(user, User), 'Argument of wrong type!'
    db = connect()
    return db['users'].insert_one(User.__dict__)


def delete_user(user):
    assert isinstance(user, User), 'Argument of wrong type!'
    db = connect()
    return db['users'].delete_one({'username': user.username})


def check_credentials(creds):
    assert isinstance(creds, Credentials), 'Argument of wrong type!'
    db = connect()
    user = db['users'].find_one({
        'username': creds.username,
        'password': creds.password
    })
    if (user != None):
        user = User.fromJson(user)
    return user


def init():
    if (not find_user("admin")):
        admin = User("admin", privivileged=True)
        insert_user(admin)
    else:
        print("Database already initialized")
